velox/common/memory/MappedMemory.cpp

```diff
@@ -180,42 +218,43 @@ bool MappedMemoryImpl::allocate(
   throw std::runtime_error("Not implemented");
 }
 
-MachinePageCount MappedMemoryImpl::allocationSize(
+bool MappedMemoryImpl::allocateContiguous(
     MachinePageCount numPages,
-    MachinePageCount minSizeClass,
-    std::array<int32_t, kMaxSizeClasses>* sizeIndices,
-    std::array<int32_t, kMaxSizeClasses>* sizeCounts,
-    int32_t* numSizes) const {
-  int32_t needed = numPages;
-  int32_t pagesToAlloc = 0;
-  *numSizes = 0;
-  VELOX_CHECK_LE(
-      minSizeClass,
-      sizes_.back(),
-      "Requesting minimum size larger than largest size class");
-  for (int32_t sizeIndex = sizes_.size() - 1; sizeIndex >= 0; sizeIndex--) {
-    int32_t size = sizes_[sizeIndex];
-    bool isSmallest = sizeIndex == 0 || sizes_[sizeIndex - 1] < minSizeClass;
-    // If the size is less than 1/8 of the size from the next larger,
-    // use the next larger size.
-    if (size > (needed + (needed / 8)) && !isSmallest) {
-      continue;
-    }
-    int32_t numUnits = std::max(1, needed / size);
-    needed -= numUnits * size;
-    if (isSmallest && needed > 0) {
-      // If needed / size had a remainder, add one unit of smallest class.
-      numUnits++;
-      needed -= size;
+    Allocation* FOLLY_NULLABLE collateral,
+    ContiguousAllocation& allocation,
+    std::function<void(int64_t)> beforeAllocCB) {
+  MachinePageCount collateralSize = 0;
+  if (collateral) {
+    collateralSize = free(*collateral) / kPageSize;
+  }
+  auto largeCollateralSize = allocation.numPages();
+  if (largeCollateralSize) {
+    if (munmap(allocation.data(), allocation.size()) < 0) {
```

Comment:
It would be nice if this could call freeContiguous() instead of copy-paste the code. It is not clear to me whether numMapped_ is being updated correctly. Is it?